agent/consul/server.go

```diff
@@ -1599,3 +1629,102 @@ will be deleted.
 
 Please see https://www.consul.io/docs/guides/outage.html for more information.
 `
+
+type RequestRecorder struct {
+	Logger      hclog.Logger
+	Tracer      trace.Tracer
+	SpanOptions []trace.SpanStartOption
+}
+
+func (r *RequestRecorder) Record(requestName string, rpcType string, start time.Time, request interface{}, respErrored bool) {
+
+	// TODO(FFMMM): consider turning errored into 1/0
+	errored := "false"
+	if respErrored {
+		errored = "true"
+	}
+	reqType := requestType(request)
+
+	labels := []metrics.Label{
+		{Name: "method", Value: requestName},
+		{Name: "errored", Value: errored},
+		{Name: "request_type", Value: reqType},
+		{Name: "rpc_type", Value: rpcType},
+	}
+
+	// TODO(FFMMM): consider using Telemetry API call here
+	metrics.MeasureSinceWithLabels([]string{"rpc", "request"}, start, labels)
+
+	r.Logger.Debug("request recorder",
+		"method", requestName,
+		"errored", errored,
+		"request_type", reqType,
+		"rpc_type", rpcType,
+		"elapsed", time.Since(start))
+}
+
+func (r *RequestRecorder) RecordSpan(ctx context.Context, requestName string, rpcType string, request interface{}, respError error) {
+	var respErrored = false
+	if respError != nil {
+		respErrored = true
+	}
+	span := trace.SpanFromContext(ctx)
+	reqType := requestType(request)
+
+	info := rpcinfo(request)
+	if info != nil {
+		span.SetAttributes(
+			attribute.Key("rpc.allow_stale").Bool(info.AllowStaleRead()),
+		)
+	} else {
+		r.Logger.Debug("failed to get RPCInfo from request in record span", "request-name", requestName, "rpc-type", rpcType)
+	}
+
+	span.SetAttributes(
+		attribute.Key("rpc.method").String(requestName),
+		attribute.Key("rpc.errored").Bool(respErrored),
+		attribute.Key("rpc.request_type").String(reqType),
+		attribute.Key("rpc.type").String(rpcType),
+	)
+}
+
+func rpcinfo(req interface{}) structs.RPCInfo {
+	if r, ok := req.(structs.RPCInfo); ok {
+		return r
+	}
+	return nil
+}
+
+func requestType(req interface{}) string {
+	if r, ok := req.(interface{ IsRead() bool }); ok && r.IsRead() {
+		return "read"
+	}
+	return "write"
+}
+
+func GetNetRPCInterceptor(recorder RequestRecorder) rpc.ServerServiceCallInterceptor {
+	return func(reqServiceMethod string, argv, replyv reflect.Value, handler func() error) {
+		reqStart := time.Now()
+		ctx := context.Background()
+		request := argv.Interface()
+		const rpcType = "net/rpc"
+		rpcInfo := rpcinfo(request)
+		if rpcInfo != nil {
+			carrier := rpcInfo.GetMetadata()
+			ctx = otel.GetTextMapPropagator().Extract(ctx, propagation.MapCarrier(carrier))
+		} else {
+			recorder.Logger.Debug("failed to get metadata to extract carrier", "request-name", reqServiceMethod, "rpc-type", rpcType)
+		}
+		ctx, span := recorder.Tracer.Start(ctx, "Consul net/rpc", recorder.SpanOptions...)
+		defer span.End()
+
+		err := handler()
```

Comment:
Pull out our context carrier from RPCInfo and use it to start a span. This will link it with the parent span context if it exists